File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/StreamStatusTraceEmitterIterator.java
package io.airbyte.integrations.source.mysql;

import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility;
import io.airbyte.commons.stream.AirbyteStreamStatusHolder;
import io.airbyte.commons.util.AutoCloseableIterator;
import io.airbyte.protocol.models.v0.AirbyteMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class StreamStatusTraceEmitterIterator implements AutoCloseableIterator<AirbyteMessage> {
    private static final Logger LOGGER = LoggerFactory.getLogger(StreamStatusTraceEmitterIterator.class);
    private boolean emitted = false;
    private AirbyteStreamStatusHolder statusHolder;

    public StreamStatusTraceEmitterIterator(final AirbyteStreamStatusHolder statusHolder) {
        this.statusHolder = statusHolder;
    }
    @Override
    public boolean hasNext() {
        return !emitted;
    }

    @Override
    public AirbyteMessage next() {
        emitted = true;
        return AirbyteTraceMessageUtility.INSTANCE.makeStreamStatusTraceAirbyteMessage(statusHolder);
    }

    @Override
    public void close() throws Exception {
        // no-op
    }

    @Override
    public void remove() {
        //no-op
    }
}

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/streamstatus/StreamStatusTraceEmitterIterator.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.source.relationaldb.streamstatus

import io.airbyte.cdk.integrations.base.AirbyteTraceMessageUtility.makeStreamStatusTraceAirbyteMessage
import io.airbyte.commons.stream.AirbyteStreamStatusHolder
import io.airbyte.commons.util.AutoCloseableIterator
import io.airbyte.protocol.models.v0.AirbyteMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class StreamStatusTraceEmitterIterator(private val statusHolder: AirbyteStreamStatusHolder) :
    AutoCloseableIterator<AirbyteMessage?> {
    private var emitted = false

    override fun hasNext(): Boolean {
        return !emitted
    }

    override fun next(): AirbyteMessage {
        emitted = true
        return makeStreamStatusTraceAirbyteMessage(statusHolder)
    }

    @Throws(Exception::class)
    override fun close() {
        // no-op
    }

    override fun remove() {
        // no-op
    }

    companion object {
        private val LOGGER: Logger =
            LoggerFactory.getLogger(StreamStatusTraceEmitterIterator::class.java)
    }
}
